File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.py
def solution(dartResult):
    score = []
    tmp = 0
    for i in range(len(dartResult)):
        v = dartResult[i]
        if v.isdigit():
            if v == "1":
                if dartResult[i + 1] == "0":
                    tmp = 10
                    i += 1
                    continue
            tmp += int(v)
        elif v in "SDT":
            if v == "D":
                tmp **= 2
            elif v == "T":
                tmp **= 3
            score.append(tmp)
            tmp = 0
        else:
            if v == "*":
                score[-1] *= 2
                if len(score) > 1:
                    score[-2] *= 2
            else:
                score[-1] *= -1
    return sum(score)

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.java
import java.util.*;

class Solution {
    public int solution(String dartResult) {
        List<Integer> result = new ArrayList<>();

        StringBuilder sb = new StringBuilder();
        
        for (char c : dartResult.toCharArray()) {
            if (Character.isDigit(c)) {
                sb.append(c);
                continue;
            }

            
            if (c == 'S' || c == 'D' || c == 'T') {
                int v = Integer.parseInt(sb.toString());
                sb.setLength(0);

                if (c == 'D') v = v*v;
                else if (c == 'T') v = v*v*v;
                
                result.add(v);
                continue;
            }
        
            if (c == '*') {
                int size = result.size();
                result.set(size-1, result.get(size-1)*2);
                if (size > 1) result.set(size-2, result.get(size-2)*2);
            }
            
            if (c == '#') {
                int size = result.size();
                result.set(size-1, result.get(size-1)*-1);
            }
        }

        return result.stream().mapToInt(v -> v).sum();
    }
}